In [2]:
#In [1]
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
# Load dataset Iris dari scikit-learn
iris = load_iris()

### In [1]: Persiapan Library dan Load Dataset
Sel ini memanggil library numpy untuk komputasi angka, pandas untuk membuat tabel data (DataFrame). Kemudian mengunduh dataset bunga Iris bawaan dari library scikit-learn.

In [3]:
# In [2]
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]
# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]
print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


### In [2]: Membuat Tabel (DataFrame) dan Menentukan Fitur (X) & Target (y)
Sel ini membuat Dataset mentah diubah menjadi tabel yang rapi dengan kolom berisi nama fitur (sepal length, sepal width, petal length, petal width).Lalu X (Fitur): Adalah data fisik bunga (tanpa kolom spesies) serta y (Target): Adalah kolom "Species" (setosa, versicolor, virginica). 

In [4]:
# In [3]
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

### In [3]: Label Encoding (Mengubah Teks menjadi Angka)
Fungsi LabelEncoder mengubah nama spesies dari teks menjadi angka indeks: setosa = 0, versicolor = 1, dan virginica = 2.

In [5]:
# In [4]
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

### In [4]: Membagi Data Latih (Train) dan Data Uji (Test)
Total dataset Iris ada 150 data. Fungsi ini membelah data menjadi dua bagian:

80% (Data Latih): Digunakan untuk melatih algoritma agar pintar.

20% (Data Uji): Digunakan sebagai soal ujian untuk mengetes kepintaran algoritma tadi.

In [6]:
# In [5]
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

### In [5]: Membuat dan Melatih Model Gaussian Naïve Bayes
Menggunakan GaussianNB karena data berupa angka desimal berkelanjutan (kontinu). Fungsi .fit(X_train, y_train) adalah proses di mana mesin "belajar" mencocokkan ciri-ciri fisik bunga dengan spesiesnya.

In [7]:
# In [6]
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
NB_model = model.fit(X_train, y_train)

### In [6]: Menguji Model dan Melihat Akurasi
Meminta model memprediksi spesies dari Data Uji (X_test), lalu membandingkan hasil tebakan model dengan kunci jawaban aslinya (y_test). Hasilnya akan muncul 96.66%, artinya model akurat.

In [8]:
# In [7]
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

### In [7]: Memprediksi Data Baru
Membungkus angka tersebut menjadi tabel pd.DataFrame dengan nama kolom yang sama seperti saat model dilatih. Fungsi inverse_transform mengubah kembali angka tebakan model (0,1,2) menjadi teks nama spesies ('setosa').

In [9]:
# In [8]
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


### In [8]: Evaluasi Akurasi Model
Membandingkan hasil tebakan model dengan kunci jawaban aslinya (y_test) menggunakan accuracy_score. Hasilnya diperoleh akurasi 96.67%.

In [10]:
# In [9]
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


C:\Users\ASUS TUF\AppData\Roaming\Python\Python314\site-packages\sklearn\utils\validation.py:2830: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


### In [9]: Prediksi Data Baru Menggunakan Format List / NumPy Array (Penyebab Warning)
sel ini mencoba memprediksi data baru `[5.1, 3.5, 1.4, 0.2]` secara langsung menggunakan format list/array. 
Meskipun hasil prediksi keluar (`['setosa']`), sistem memunculkan pesan **UserWarning**. Hal ini terjadi karena model `GaussianNB` sebelumnya dilatih menggunakan DataFrame yang memiliki nama fitur/kolom (`iris.feature_names`), sedangkan input data ini dikirim tanpa nama kolom.

In [11]:
# In [10]
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


### In [10]: Solusi Prediksi Data Baru Menggunakan DataFrame (Bebas Warning)
Sebagai solusi dari peringatan pada `In [9]`, data baru diubah terlebih dahulu menjadi objek `DataFrame` lengkap dengan nama kolom aslinya (`columns=iris.feature_names`). 
Dengan struktur data yang konsisten sesuai data saat pelatihan model, prediksi menghasilkan output `['setosa']` secara bersih tanpa memunculkan peringatan (*warning*).

In [14]:
#TUGAS
# Membuat array numpy dari data baru
data_baru = np.array([[7.3, 3.8, 4.6, 2.3]])

# Mengubah menjadi DataFrame agar tidak muncul peringatan (Warning) karena missing feature names
input_data = pd.DataFrame(data_baru, columns=iris.feature_names)

# Melakukan prediksi
prediksi_baru = NB_model.predict(input_data)

# Mengembalikan nilai angka (0,1,2) ke nama spesies aslinya (setosa, versicolor, virginica)
prediction_label = lab.inverse_transform(prediksi_baru)

# Mencetak hasil
print(f"Data input baru: {data_baru[0]}")
print(f"Hasil prediksi untuk input baru adalah: {prediction_label}")

Data input baru: [7.3 3.8 4.6 2.3]
Hasil prediksi untuk input baru adalah: ['virginica']


### Kesimpulan Hasil Prediksi Tugas:
Berdasarkan perhitungan probabilitas algoritma **Gaussian Naïve Bayes**, data baru dengan ukuran kelopak dan mahkota `[7.3, 3.8, 4.6, 2.3]` diprediksi masuk ke dalam spesies **`virginica`**. 

Hal ini logis secara biologis/data karena nilai *sepal length* (7.3) dan *petal width* (2.3) tergolong sangat besar, yang merupakan karakteristik utama dari spesies Iris Virginica dibandingkan Setosa maupun Versicolor.